client/peerset/src/protocol_controller.rs

```diff
@@ -0,0 +1,559 @@
+						Direction::Outbound => self.num_out -= 1,
+					}
+					self.drop_connection(peer_id);
+				},
+				PeerState::NotConnected => {},
+			},
+			None => {
+				trace!(
+					target: "peerset",
+					"Trying to remove unknown peer {} from {:?}",
+					peer_id, self.set_id,
+				);
+			},
+		}
+	}
+
+	fn on_incoming_connection(&mut self, peer_id: PeerId, incoming_index: IncomingIndex) {
+		if self.reserved_only && !self.reserved_nodes.contains_key(&peer_id) {
+			self.reject_connection(incoming_index);
+			return
+		}
+
+		// Check if the node is reserved first.
+		if let Some(state) = self.reserved_nodes.get_mut(&peer_id) {
+			match state {
+				// If we're already connected, don't answer, as the docs mention.
+				PeerState::Connected(_) => {},
+				PeerState::NotConnected => {
+					// It's questionable whether we should check a reputation of reserved node.
+					// FIXME: unable to call `self.is_banned()` because of borrowed `self`.
+					if self.peerset_handle.is_banned(peer_id) {
+						self.reject_connection(incoming_index);
+					} else {
+						*state = PeerState::Connected(Direction::Inbound);
+						self.accept_connection(incoming_index);
+					}
+				},
+			}
+			return
+		}
+
+		// Adding incoming peer to our set of peers even before we decide whether to accept
+		// it is questionable, but this is how it was implemented in the original `Peerset`.
+		let state = self.nodes.entry(peer_id).or_insert(PeerState::NotConnected);
+		match state {
+			// If we're already connected, don't answer, as the docs mention.
+			PeerState::Connected(_) => return,
+			PeerState::NotConnected => {
+				// FIXME: unable to call `self.is_banned()` because of borrowed `self`.
+				if self.peerset_handle.is_banned(peer_id) {
+					self.reject_connection(incoming_index);
+					return
+				}
+
+				if self.num_in >= self.max_in {
+					self.reject_connection(incoming_index);
+					return
+				}
+
+				self.num_in += 1;
+				*state = PeerState::Connected(Direction::Inbound);
+				self.accept_connection(incoming_index);
+			},
+		}
+	}
+
+	/// Returns Err(PeerId) if peer wasn't connected.
+	fn on_peer_dropped(&mut self, peer_id: PeerId, reason: DropReason) -> Result<(), PeerId> {
+		let mut is_reserved = true;
+		let state = self
+			.reserved_nodes
+			.get_mut(&peer_id)
+			.or_else(|| {
+				is_reserved = false;
+				self.nodes.get_mut(&peer_id)
+			})
+			.ok_or(peer_id)?;
+		match state {
+			PeerState::Connected(d) => {
+				if is_reserved {
+					*state = PeerState::NotConnected;
+				} else {
+					match d {
+						Direction::Inbound => self.num_in -= 1,
+						Direction::Outbound => self.num_out -= 1,
+					}
+					if let DropReason::Refused = reason {
+						self.nodes.remove(&peer_id);
+					} else {
+						*state = PeerState::NotConnected;
+					}
+				}
+				self.report_disconnect(peer_id);
+				Ok(())
+			},
+			PeerState::NotConnected => Err(peer_id),
+		}
+	}
+
+	fn alloc_slots(&mut self) {
+		if self.num_out >= self.max_out {
```

Comment:
`num_out` should never larger than `max_out` since reserved peers are no longer counted in `num_out`

Reply:
Good point but in that case the code shouldn't return here but cull some connections if `num_out` is `max_out` because the number of occupied slots is more than the protocol specified

Reply:
I also though about this, but decided to keep the implementation the same as it is in `Peerset`.

Reply:
> BTW, I just realized that `alloc_slots` in the current implementation is incompatible with the previous behaviour. Here I ask `Peerset` for the candidates, but originally they were selected from the known peers on the specific peer set (different for each protocol). This will require changing the interface between `ProtocolController` and `Peerset`.

I think this is fine. We're introducing breaking changes anyway throughout the connection handling logic. `Peerset` should know which protocols are supported by each peer so when a protocol queries new suitable peers, `Peerset` would return those peers that support the protocol. The information of which protocols are supported are learned through `Identify`. As long as the tests show that we're able to maintain the desired number of inbound and outbound connections and that the protocol attempts to establish connections to new peers in decreasing order of their reputations, I think it's good enough for now. We can improve the peer selection later on.